src/bookshelf/src/bksw_sec.erl

```diff
@@ -150,26 +179,89 @@ encode_sign_error_response(AccessKeyId, Signature,
     Req2 = wrq:set_resp_body(Body, Req1),
     {{halt, 403}, Req2, Context}.
 
-encode_access_denied_error_response(RequestId, Req0, Context) ->
-    Req1 = bksw_req:with_amz_id_2(Req0),
-    Body = bksw_xml:access_denied_error(RequestId),
-    Req2 = wrq:set_resp_body(Body, Req1),
-    {{halt, 403}, Req2, Context}.
+% split  "<bucketname>/<key>" (possibly leading and/or trailing /) into {"bucketname", "key"}
+% Path = "<bucketname>/<key>"
+-spec get_bucket_key(Path::string()) -> {string(), string()}.
+get_bucket_key(Path) ->
+    case string:lexemes(Path, "/") of
+        [            ] -> {"",     ""};
+        [Bucket      ] -> {Bucket, ""};
+        [Bucket | Key] -> {Bucket, filename:join(Key)}
+    end.
 
-split_authorization([$A, $W, $S, $\s, $: | Rest]) ->
-    [<<>>, Rest];
-split_authorization([$A, $W, $S, $\s  | Rest]) ->
-    string:tokens(Rest, ":").
+% get host and toggle the port (add port or remove it)
+-spec get_host_toggleport(string(), aws_config()) -> string().
+get_host_toggleport(Host, Config) ->
```

Comment:
Since we're using it in most cases, maybe assign S3Port in the function header instead of dereferencing in each case.